File: leetcode/BackTracking/leetcode-79/79.go
package backtracking

func exist(board [][]byte, word string) bool {
	if board == nil || len(board) == 0 || len(board[0]) == 0 || len(word) == 0 {
		return false
	}

	for i := 0; i < len(board); i++ {
		for j := 0; j < len(board[i]); j++ {
			if helper(board, i, j, word, 0) {
				return true
			}
		}
	}
	return false
}

func helper(board [][]byte, row, col int, word string, loc int) bool {
	if loc == len(word) {
		return true
	}

	if row < 0 || row >= len(board) || col < 0 || col >= len(board[0]) {
		return false
	}

	if board[row][col] == '0' {
		return false
	}

	if board[row][col] != word[loc] {
		return false
	}

	c := board[row][col]
	board[row][col] = '0'
	res := helper(board, row, col+1, word, loc+1) || helper(board, row, col-1, word, loc+1) || helper(board, row+1, col, word, loc+1) || helper(board, row-1, col, word, loc+1)

	if res == false {
		board[row][col] = c
	}

	return res
}

File: leetcode/BackTracking/leetcode-79/79.java
class Solution {
    public boolean exist(char[][] board, String word) {
        if (board == null || board.length == 0 || board[0].length == 0 || word == null) {
            return false;
        }

        int row = board.length;
        int col = board[0].length;
        boolean[][] flag = new boolean[row][col];
        for (int i = 0; i < row; i++) {
            for (int j = 0; j < col; j++) {
                flag[i][j] = false;
            }
        }

        for (int i = 0; i < row; i++) {
            for (int j = 0; j < col; j++) {
                if (helper(board, i, j, flag, word, 0)) {
                    return true;
                }
            }
        }

        return false;
    }

    private boolean helper(char[][] board, int row, int col, boolean[][] flag, String word, int loc) {
        if (loc == word.length()) {
            return true;
        }

        if (row < 0 || row >= board.length || col < 0 || col >= board[0].length) {
            return false;
        }

        if (flag[row][col] == true) {
            return false;
        }

        if (board[row][col] != word.charAt(loc)) {
            return false;
        }
        
        flag[row][col] = true;
        int[][] next = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
        for (int i = 0; i < 4; i++) {
            if (helper(board, row + next[i][0], col + next[i][1], flag, word, loc + 1)) {
                return true;
            }
        }
        flag[row][col] = false;
        return false;
    }
}